Kernel/FileSystem/FATFileSystem.cpp

```diff
@@ -0,0 +1,352 @@
+            filename.append(lfn_entry.characters2[2]);
+            filename.append(lfn_entry.characters2[3]);
+            filename.append(lfn_entry.characters2[4]);
+            filename.append(lfn_entry.characters2[5]);
+            filename.append(lfn_entry.characters3[0]);
+            filename.append(lfn_entry.characters3[1]);
+        }
+
+        return TRY(KString::try_create(byte_terminated_string(filename.string_view(), 0xFF)));
+    }
+
+    VERIFY_NOT_REACHED();
+}
+
+time_t FATInode::fat_date_time(u16 date, u16 time) const
+{
+    time_t year = (date >> 9) + (1980 - 1970);
+    time_t month = (date >> 5) & 0b1111;
+    time_t day = date & 0b11111;
+
+    time_t hour = time >> 11;
+    time_t minute = (time >> 5) & 0b1111111;
+    time_t second = (time & 0b11111) << 1;
+
+    return second + minute * 60 + hour * 3600
+        + day + (24 * 3600) + month * (30 * 24 * 3600) + year * (365 * 24 * 3600);
+}
+
+// TODO: This might be worth putting as a part of the StringView class
+StringView FATInode::byte_terminated_string(StringView string, u8 fill_byte)
+{
+    while (string.length() > 0) {
+        if (string[string.length() - 1] != fill_byte)
+            break;
+        string = string.substring_view(0, string.length() - 1);
+    }
+
+    return string;
+}
+
+ErrorOr<size_t> FATInode::read_bytes_locked(off_t offset, size_t size, UserOrKernelBuffer& buffer, OpenFileDescription*) const
+{
+    dbgln_if(FAT_DEBUG, "FATFS: Reading inode {}: size: {} offset: {}", identifier().index(), size, offset);
+
+    // FIXME: Read only the needed blocks instead of the whole file
+    auto blocks = TRY(const_cast<FATInode&>(*this).read_block_list());
+    TRY(buffer.write(blocks->data() + offset, min(size, m_block_list.size() * fs().m_logical_block_size - offset)));
+
+    return min(size, m_block_list.size() * fs().m_logical_block_size - offset);
+}
+
+InodeMetadata FATInode::metadata() const
+{
+    return m_metadata;
+}
+
+ErrorOr<void> FATInode::traverse_as_directory(Function<ErrorOr<void>(FileSystem::DirectoryEntryView const&)> callback) const
+{
+    MutexLocker locker(m_inode_lock);
+
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    Vector<RefPtr<FATInode>> children = TRY(const_cast<FATInode&>(*this).traverse());
+
+    for (auto const& inode : children) {
+        TRY(callback({ inode->m_filename->view(), inode->identifier(), static_cast<u8>(inode->m_entry.attributes) }));
+    }
+
+    return {};
+}
+
+ErrorOr<NonnullLockRefPtr<Inode>> FATInode::lookup(StringView name)
+{
+    MutexLocker locker(m_inode_lock);
+
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    Vector<RefPtr<FATInode>> children = TRY(traverse());
+
+    for (auto& inode : children) {
+        if (strncmp(reinterpret_cast<char const*>(inode->m_filename->bytes().data()), reinterpret_cast<char const*>(name.bytes().data()), name.length()) == 0) {
+            return TRY(adopt_nonnull_lock_ref_or_enomem(inode.leak_ref()));
```

Comment:
This looks very suspicious.

Why are we leaking a reference? 

This makes it look like we have some places holding a RefPtr to these FATInode objects, and other places holding a Locking RefPtr. That seems very unsafe. If the parent INode classes expect things to be returned as locking RefPtrs, we should probably use that everywhere to not trip over ourselves later and fix the locking story at a higher level.